aggregate_root/lib/aggregate_root.rb

```diff
@@ -26,13 +26,13 @@ def store(stream_name = loaded_from_stream_name, event_store: default_event_stor
     @unpublished_events = nil
   end
 
-  private
-  attr_reader :loaded_from_stream_name
-
   def unpublished_events
     @unpublished_events ||= []
```

Comment:
Would it be safer to return a clone of this list for the public accessor? I'm worried that debugging of situations when someone mutates this object will be difficult.

Reply:
yes, agree... I would keep original method private. Maybe returning iterator function will be better and much safer. And will not require always calling clone even for internal usage.

Reply:
for testing you can always use metaprograming:
```
require 'rspec/expectations'
require_relative 'utils'

RSpec::Matchers.define :publish do |expected_events|
  match do |aggregate|
    @stripped_actual   = as_string(aggregate.__send__("unpublished_events"))
    @stripped_expected = as_string(expected_events)
    expect(@stripped_actual).to eq @stripped_expected
  end

  failure_message do |aggregate|
    message = "Expected aggregate:\n\n#{aggregate.inspect}\n\nto " \
              "publish events:\n\n#{@stripped_actual}\n\nwould match:\n\n#{@stripped_expected}"
    message += "\n\nDiff: #{differ.diff_as_string(@stripped_actual, @stripped_expected)}"
    message
  end

  def differ
    RSpec::Support::Differ.new(
      color: RSpec::Matchers.configuration.color?
    )
  end
end
```
Its not perfect but I'm fine with using little bit of "magic". Exposing this method to the public will open way of overusing it or just misuse of aggregate. I know from past experiences that people even thinking about using aggregate root state as public and skip making read models, I don't think that giving here to much freedom is a best way to deal with testing problem